BUG: Fix numpy boolean subtraction error in Series.diff

Unprocessable opened the pull request:
A new version of the pull request (#27755), the other one was a bit behind. Original text below.

This fixes #17294, for more than three years now NumPy has not allowed the subtraction of boolean series. 

TypeError                                 Traceback (most recent call last)
<ipython-input-46-3da3b949c6bd> in <module>
      1 data = pd.Series([0,-1,-2,-3,-4,-3,-2,-1,0,-1,-1,0,-1,-2,-3,-2,0])
      2 filtered = data.between(-2,0, inclusive = True)
----> 3 filtered.diff()
      4 print(filtered)

~\AppData\Local\Continuum\anaconda3\lib\site-packages\pandas\core\series.py in diff(self, periods)
   2191         dtype: float64
   2192         """
-> 2193         result = algorithms.diff(com.values_from_object(self), periods)
   2194         return self._constructor(result, index=self.index).__finalize__(self)
   2195 

~\AppData\Local\Continuum\anaconda3\lib\site-packages\pandas\core\algorithms.py in diff(arr, n, axis)
   1817             out_arr[res_indexer] = result
   1818         else:
-> 1819             out_arr[res_indexer] = arr[res_indexer] - arr[lag_indexer]
   1820 
   1821     if is_timedelta:

TypeError: numpy boolean subtract, the `-` operator, is deprecated, use the bitwise_xor, the `^` operator, or the logical_xor function instead.

- [x] closes #17294
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Unprocessable committed:
Fix numpy boolean subtraction error in Series.diff

Numpy no longer allows subtraction of boolean values. Not sure if Numpy version checking should be done before this code...
Note: I haven't actually run this code, feel free to check it, but should be correct.

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-46-3da3b949c6bd> in <module>
      1 data = pd.Series([0,-1,-2,-3,-4,-3,-2,-1,0,-1,-1,0,-1,-2,-3,-2,0])
      2 filtered = data.between(-2,0, inclusive = True)
----> 3 filtered.diff()
      4 print(filtered)

~\AppData\Local\Continuum\anaconda3\lib\site-packages\pandas\core\series.py in diff(self, periods)
   2191         dtype: float64
   2192         """
-> 2193         result = algorithms.diff(com.values_from_object(self), periods)
   2194         return self._constructor(result, index=self.index).__finalize__(self)
   2195 

~\AppData\Local\Continuum\anaconda3\lib\site-packages\pandas\core\algorithms.py in diff(arr, n, axis)
   1817             out_arr[res_indexer] = result
   1818         else:
-> 1819             out_arr[res_indexer] = arr[res_indexer] - arr[lag_indexer]
   1820 
   1821     if is_timedelta:

TypeError: numpy boolean subtract, the `-` operator, is deprecated, use the bitwise_xor, the `^` operator, or the logical_xor function instead.

Unprocessable committed:
Added is_bool outside if function

Unprocessable committed:
Added test_diff function

Unprocessable committed:
Update test_diff.py

Unprocessable committed:
Changing style to match PEP

Unprocessable committed:
Update test_diff.py

Unprocessable committed:
Using tm to assert series differences

Unprocessable committed:
Added diff bug description to changelog

Unprocessable committed:
Added issue number

Unprocessable committed:
Adding timeseries tests as well

Unprocessable committed:
Removed test_diff, it is added to test_diff.py

I copied the data over to the other file, diff tests should be in one file and the boolean test cannot be added to this one as it is not a time series.

Unprocessable committed:
Removed too much code, importing nan again

Unprocessable committed:
STYLE: formatting...

Unprocessable committed:
Splitting into two functions

Unprocessable committed:
Split into 3 functions

Unprocessable committed:
Update test_diff.py

Unprocessable committed:
Adding nan test

Unprocessable committed:
STYLE: Black formatting

Unprocessable committed:
Import order changed

Unprocessable committed:
Update test_diff.py

Unprocessable committed:
Update test_diff.py

Unprocessable committed:
Moving test_diff.py to test_analytics.py

As requested...

Unprocessable committed:
Removed test_diff.py, added to test_analytics

Unprocessable committed:
Removing tailing whitespaces

Unprocessable committed:
STYLE: Black formatting

Unprocessable committed:
Update test_analytics.py

Unprocessable committed:
Update test_analytics.py

Unprocessable committed:
Removing self.ts from code

Unprocessable committed:
Update test_analytics.py

Unprocessable committed:
Adding tm.makeTimeSeries, as in the original

Unprocessable committed:
STYLE: Black

Commit:
Adding updates

Commit:
BUG: Fix numpy boolean subtraction error in Series.diff (#27755)

Commit:
Merge branch 'patch-2' of https://github.com/Unprocessable/pandas into patch-2

Commit:
Resolved merge conflict added changes to older whatsnew.

Unprocessable committed:
Update v0.25.2.rst

Adding fix to Other

Commit:
Merge branch 'patch-2' of https://github.com/Unprocessable/pandas into patch-2

Unprocessable committed:
Update v0.25.1.rst

simonjayhawkins commented on doc/source/whatsnew/v0.25.2.rst:
```diff
@@ -100,7 +100,7 @@ Other
 
 - Compatibility with Python 3.8 in :meth:`DataFrame.query` (:issue:`27261`)
--
+- Bug in :meth:`Series.diff` where a boolean series would cause a TypeError (the - operator is deprecated) when using NumPy >= 0.13.0 (:issue:`17294`)
```
can you move this to 1.0.0.

jreback commented on pandas/tests/series/test_analytics.py:
```diff
@@ -237,6 +238,59 @@ def test_npdiff(self):
+        # boolean nan series
+        s = Series([False, True, nan, False, False])
+        result = s.diff()
+        assert_series_equal(result, Series([nan, 1, nan, nan, 0], dtype="object"))
```
can you write the expected on a separate line rather than inlining, e.g.

result = 
expected = 
assert_series(result, expected)

is the pattern

jreback commented on pandas/tests/series/test_analytics.py:
```diff
@@ -237,6 +238,59 @@ def test_npdiff(self):
         r = np.diff(s)
         assert_series_equal(Series([nan, 0, 0, 0, nan]), r)
 
+    def test_dt_nm_bool_diff(self):
```
name this test_diff

jreback commented on pandas/tests/series/test_analytics.py:
```diff
@@ -237,6 +238,59 @@ def test_npdiff(self):
+            result, Series(TimedeltaIndex(["NaT"] + ["1 days"] * 4), name="foo")
+        )
+
+        # boolean series
```
add the issue number as a comment

jreback commented on pandas/tests/series/test_analytics.py:
```diff
@@ -237,6 +238,59 @@ def test_npdiff(self):
         r = np.diff(s)
         assert_series_equal(Series([nan, 0, 0, 0, nan]), r)
 
+    def test_dt_nm_bool_diff(self):
```
bonus point if you can parameterize this test

Unprocessable replied:
Thanks for the offer, but for now this will have to do. I basically added 3 lines of code that should have been added more than a few years ago. The need for writing a few tests I can get behind, but I would rather not completely refactor the whole file. The rest is implemented.

Commit:
Refactored code and added version number

Commit:
Merge branch 'patch-2' of https://github.com/Unprocessable/pandas into patch-2

pep8speaks commented:
Hello @Unprocessable! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`pandas/tests/series/test_analytics.py`](https://github.com/pandas-dev/pandas/blob/a174239c1c2e521a66a631020b7ee86a5223a88a/pandas/tests/series/test_analytics.py):

> [Line 281:9](https://github.com/pandas-dev/pandas/blob/a174239c1c2e521a66a631020b7ee86a5223a88a/pandas/tests/series/test_analytics.py#L281): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent

Commit:
Fixing typo

Commit:
Moving comment

Commit:
Removed a space

Commit:
Fixed typo

TomAugspurger commented on doc/source/whatsnew/v1.0.0.rst:
```diff
@@ -211,6 +211,7 @@ Other
 - Trying to set the ``display.precision``, ``display.max_rows`` or ``display.max_columns`` using :meth:`set_option` to anything but a ``None`` or a positive int will raise a ``ValueError`` (:issue:`23348`)
 - Using :meth:`DataFrame.replace` with overlapping keys in a nested dictionary will no longer raise, now matching the behavior of a flat dictionary (:issue:`27660`)
 - :meth:`DataFrame.to_csv` and :meth:`Series.to_csv` now support dicts as ``compression`` argument with key ``'method'`` being the compression method and others as additional compression options when the compression method is ``'zip'``. (:issue:`26023`)
+- Bug in :meth:`Series.diff` where a boolean series would cause a TypeError (the - operator is deprecated) when using NumPy >= 0.13.0 (:issue:`17294`)
```
Should this be NumPy 1.13.0?

Unprocessable replied:
You are right! The bug would have been there for a very very long time otherwise. Should be fixed now.

Unprocessable replied:
Before that version it would show a deprecation warning, no idea for how long NumPy did that.

Commit:
Added spaces

Commit:
Fixed typo

TomAugspurger commented on pandas/tests/series/test_analytics.py:
```diff
@@ -237,6 +238,60 @@ def test_npdiff(self):
+        expected = Series([nan, True, False, True, False])
+        assert_series_equal(result, expected)
+
+        # boolean nan series
```
I don't agree that this should differ from the previous case. IMO, both of these are boolean dtype. Pandas just doesn't have a nullable boolean dtype (yet) but we generally try to treat a mix of bool and NA the same as bool.

So I think that this should also be object dtype, but the values should be booleans instead of integers. Is that difficult to support? I suppose it'll require a scan of the values for the `object`-dtype case, to ensure that everything is boolean or NA?

Unprocessable replied:
When I previously tested my code I did notice a difference, hence why I added the tests as such. If you are sure that pandas does not treat it any different and as such does not require testing this now and in the future, I will gladly remove the first test.

In regards to your second point, I could definitely change the code to create only a boolean series instead, but why? The same could be done with logic instead of diff. I think the point of diff is to see the difference between the current value and the previous one (e.g. True - False = 1 and False - True = -1). This is also in line with the subtraction done by diff on other non-boolean series and what happens if you subtract booleans in Python. I found this bug looking for ways to detect these changes. Or is that not what you meant?

TomAugspurger replied:
I'm having a hard time following, but IMO these two should have the same output

```python
In [3]: pd.Series([True, False, True]).diff()
Out[3]:
0     NaN
1    True
2    True
dtype: object

In [4]: pd.Series([True, False, True]).astype(object).diff()
Out[4]:
0    NaN
1     -1
2      1
dtype: object
```

I think both of those should be `Out[3]`, the same result as `np.diff` with the missing values handled appropriately.

Unprocessable replied:
I seem to be missing something, when I run np.diff on the series it results in this:

```
In [2]: np.diff(pd.Series([False, True]))
Out[2]: array([ True])

In [3]: np.diff(pd.Series([np.nan, False, True]))
Out[3]: array([nan, 1], dtype=object)
```
Which is pretty close to the current pandas implementation. Do we want to remove the first item in the list that currently results in np.nan? That would mean changing the other code in the diff function as well. Ignoring that, the results are the same as my code. As the dtypes are the problem, we could either convert the series before or do the same numpy does.

Unprocessable replied:
Just to make sure, I also tested np.diff on a numpy array with the same results:

```
In [4]: np.diff(np.array([True, False, np.nan]))
Out[4]: array([-1., nan])

In [5]: np.diff(np.array([True, False]))
Out[5]: array([ True])
```

TomAugspurger replied:
I think we want the same behavior as `np.diff` on a boolean-dtype ndarray.

Because of the missing values, we'll always return an object-dtype Series where the values are either NaN, True, or False.

cc @jbrockmendel if you have thoughts.

jbrockmendel replied:
It sounds like this would be a lot easier if we had a BoolNA in place.  Is that accurate?  Since bool-dtype is a thin wrapper over uint8, I wonder how hard it would be to adapt IntegerNA to this purpose (@WillAyd you had a PR about this a while back, right?)

WillAyd replied:
Long time ago. @jreback had something more modern in #25415

TomAugspurger replied:
Yes, this would be easier with BoolNA.

I think we want to achieve the BoolNA behavior, but with object dtype today.

jreback replied:
am -1 on doing anything with object dtype as it makes the code much too complex.

TomAugspurger commented:
Are we in agreement with https://github.com/pandas-dev/pandas/pull/28251#discussion_r320818052? That the non-NA elements should be booleans just like NumPy's `np.diff` on bool data?

WillAyd commented:
Yea I think that makes the most sense. Will take up more space but let's leave optimization as a down the road exercise to implement BoolNA

Commit:
Adding requirements

Commit:
Updating

jreback commented on doc/source/whatsnew/v1.0.0.rst:
```diff
@@ -211,6 +211,7 @@ Other
 - Trying to set the ``display.precision``, ``display.max_rows`` or ``display.max_columns`` using :meth:`set_option` to anything but a ``None`` or a positive int will raise a ``ValueError`` (:issue:`23348`)
 - Using :meth:`DataFrame.replace` with overlapping keys in a nested dictionary will no longer raise, now matching the behavior of a flat dictionary (:issue:`27660`)
 - :meth:`DataFrame.to_csv` and :meth:`Series.to_csv` now support dicts as ``compression`` argument with key ``'method'`` being the compression method and others as additional compression options when the compression method is ``'zip'``. (:issue:`26023`)
+- Bug in :meth:`Series.diff` where a boolean series would cause a TypeError (the - operator is deprecated) when using NumPy >= 1.13.0 (:issue:`17294`)
```
Bug in :meth:`Series.diff` where a boolean series would incorrectly raise ``TypeError`` (:issue:....)

Unprocessable replied:
Quick question: if people start looking for that bug ('the - operator is deprecated'), how would they know to update pandas if I remove the second part?

jreback commented on pandas/core/algorithms.py:
```diff
@@ -1923,10 +1925,15 @@ def diff(arr, n, axis=0):
     elif is_integer_dtype(dtype):
         dtype = np.float64
 
+    elif is_object_dtype(dtype):
```
I am not a fan of doing anything anything with object arrays here at all.

jreback commented on pandas/tests/series/test_analytics.py:
```diff
@@ -237,6 +238,60 @@ def test_npdiff(self):
+
+        # boolean nan series
+        s = Series([False, True, nan, False, False])
+        result = s.diff()
```
this routine is way too long can you parameterize

Unprocessable replied:
Come on people, make up your minds. I have two versions now, the original commit is the original implementation, but working with a new numpy version, the second one also dealing with np.nan object 'boolean' arrays. 

Which one should be implemented? 
@TomAugspurger and @WillAyd want the second implementation
@jreback and me the first
That leaves @jbrockmendel and @simonjayhawkins

Unprocessable replied:
@jreback Shall I rewrite the rest of those tests in pandas as well? I did not write the tests the way they are, I only added a few lines... I would rather do this in a refactor.

WillAyd replied:
Is there a particular reason why you moved this test? I think would be easier just to create another test specific to bool here and leave the time series stuff where it is

Unprocessable commented:
Come on people, make up your minds. I have two versions now, the original commit is the original implementation, but working with a new numpy version, the second one also dealing with np.nan object 'boolean' arrays.

Which one should be implemented?
@TomAugspurger and @WillAyd want the second implementation
@jreback and me the first
That leaves @jbrockmendel and @simonjayhawkins
Unless it is not a democracy, then feel free to tell me which one to pick.

jreback commented:
@Unprocessable how about waiting to update. This is slightly non-trivial as to scope.

Unprocessable commented:
@jreback You are right. I assumed that a bug that went unnoticed for several years did not affect the scope that much. I will wait, but I would like to have this bug fixed in the next release.

jreback commented:
> @jreback You are right. I assumed that a bug that went unnoticed for several years did not affect the scope that much. I will wait, but I would like to have this bug fixed in the next release if possible.

thanks @Unprocessable *many* times bugs go unfixed for long periods of time. Its only when they  are being fixed things are looked at more closely.

TomAugspurger commented:
@jreback what's your ideal behavior?

My preference is for us to implement the behavior we would desire *if* we had a nullable BoolArray. 

Do you agree that if we had BoolArray, then `Series[bool].diff() -> Series[bool]` rather than `-> Series[float]` like we do today, when there are some NAs present?

jreback commented:
> @jreback what's your ideal behavior?
> 
> My preference is for us to implement the behavior we would desire _if_ we had a nullable BoolArray.
> 
> Do you agree that if we had BoolArray, then `Series[bool].diff() -> Series[bool]` rather than `-> Series[float]` like we do today, when there are some NAs present?

yes I am not averse to implementing this, but want to do it generically. We have *very* little object support inside that routine and that's why I think adding here now is not great.

Unprocessable commented:
@TomAugspurger The original code and my first commit also do that. If the dtype is a bool, it will return a bool. The problem arises when the array or Series has the object dtype due to the presence of np.nan before running the diff function. So I would say that once the nan in the Series is seen by pandas as a bool dtype, a Series of True, False and nan will be of a bool dtype and the code will execute correctly.

TomAugspurger commented:
> yes I am not averse to implementing this, but want to do it generically. We have very little object support inside that routine and that's why I think adding here now is not great.

Do you consider this blocked by a BoolArray implementation? We'll still need to get users migrated to that API, so people with an object-dtype ndarray of bool & nan will still hit this issue.

I think for now I'm OK with supporting on object-dtype ndarray of bool & NaN, since it'll be ~5ish LoC to support properly, and it establishes the correct behavior going forward.

Unprocessable commented:
There are two separate issues:
1. diff on a boolean Series currently does not work (- deprecated)
2. the current and previous implementation caused bool + NA to return floats

My goal was definitely fixing issue 1 and maybe fixing issue 2.
@TomAugspurger if I understand you correctly, you see issue 2 as more pressing, am I correct?

TomAugspurger commented:
Fair point. Fixing 1 is probably more pressing.

But regardless of how where we fix 2, I'd like to not add tests with
"incorrect" behavior (diff returning floats) in this PR.

On Tue, Sep 10, 2019 at 9:46 AM Unprocessable <notifications@github.com>
wrote:

> There are two separate issues:
>
>    1. diff on a boolean Series currently does not work (- deprecated)
>    2. the current and previous implementation caused bool + NA to return
>    floats
>
> My goal was definitely fixing issue 1 and maybe fixing issue 2.
> @TomAugspurger <https://github.com/TomAugspurger> if I understand you
> correctly, you see issue 2 as more pressing, am I correct?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/28251?email_source=notifications&email_token=AAKAOIVJZ6V6W6V35INMD7LQI6XL5A5CNFSM4IS6G5PKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD6LLKLQ#issuecomment-529970478>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAKAOIRL6NUJ6VGD4APYKDDQI6XL5ANCNFSM4IS6G5PA>
> .
>

Unprocessable commented:
That I can get behind. So either we fix the second issue as well or we leave it for another bug report. Feel free to discuss.

Commit:
Updating documentation

Unprocessable committed:
Merge branch 'master' into patch-2

Commit:
Adding object series test

Unprocessable commented:
@TomAugspurger Do you agree that I should remove the object part, like jreback requested, so we can at least fix the first issue? I have changed the test for the nan series to test the result of an object series, so it no longer tests for incorrect behaviour.

Commit:
Merging with master

TomAugspurger commented on pandas/core/algorithms.py:
```diff
@@ -1923,10 +1925,15 @@ def diff(arr, n, axis=0):
         dtype = np.float64
 
+    elif is_object_dtype(dtype):
+        if np.all(np.isin(arr, (True, False), False) | isna(arr)):
```
This can be

```
is_obj_bool = lib.infer_dtype(arr, skipna=True) == "boolean"
```

Unprocessable replied:
I removed it, as requested.

TomAugspurger reviewed (commented):
OK, where does this stand?

My preference is to merge the smallest, correct change. So just the changes in `.diff` and the new tests that don't have any NaNs in the input.

Unprocessable committed:
Removing the extra object NaN code

Unprocessable commented:
As far as I know, this should work. It is changed back to the same code it was several tens of commits ago. If it does not work, I will fix it later. If everything checks out, feel free to merge with master.

Unprocessable reviewed (commented):
@WillAyd I did create a new test previously. There are several reasons it is like this now:

1. because @jbrockmendel [asked me to](https://github.com/pandas-dev/pandas/pull/27755#issuecomment-518289559)
2. because it does not make sense to have several tests in different files all testing diff separately
3. because the name implies a general diff test, not a time series specific one
4. because the code that is tested is all contained in the diff function, so it would be better for it be tested in a contained test, it makes debugging a failed test easier
5. the general test for diff is also included in those tests (a normal integer test), so it looks like it was never a time series specific test in the first place

I understand the need for different tests as well, so if you get a consensus of three more people agreeing it should be in different tests, I will gladly change it.

Unprocessable commented:
@TomAugspurger @jreback 
Do you currently agree that we ignore the object nan Series for now and only focus on the boolean-only series? The current commit reflects that change at the moment.

TomAugspurger commented:
Agreed with your fix in the method, and @WillAyd's comment about the test. I'd prefer a focused, new test like

```python
def test_diff_bool(self):
        s = Series([False, True, True, False, False])
        result = s.diff()
        expected = Series([nan, True, False, True, False])
        assert_series_equal(result, expected)
```

And leaving out the object-dtype Series.diff test.

Unprocessable commented:
I just tested a diff with a combination of datetime values and floats, this will crash as well. As long as there is a combination of integers, floats (nan included) or booleans, the object diff will work. While testing that specific case might not be a good idea, we do not want to introduce a patch that will make booleans from a float/int object series. For that reason I think it is better to include a test for that. 

@TomAugspurger How would you test that a diff on a combined boolean and float object series results in floats instead of booleans? I am pretty certain that someone doing a diff on [0, 1.0, True] expects [nan, -1.0, 0] instead of [nan, True, False], which is how it is written now.

TomAugspurger commented:
I'd like this diff to be as small as possible. I'd only like to fix the original issue: calling `-` on a boolean ndarray.

Let's leave the rest for another issue.

Unprocessable committed:
Adding requirements

Looks like I will be parametrizing and splitting the function into several functions. If there are no bugs, this should do.

Unprocessable committed:
BLACK formatting

Unprocessable committed:
Adding comma

Unprocessable commented:
I have been fixing the tests as well as there was a need for that from you (among others). I would rather not (re)write any tests, but I get the need for them. As the issue is now fixed and the boolean diff test is both parametrized and split from the rest, I would like this to be finished soon. 

> Let's leave the rest for another issue.
I agree with you, if anyone does not agree with how it was written, it will be possible to fix it later. For now it works, does what it should do and is definitely not worth more time nitpicking.

Unprocessable reviewed (commented):
The requests should be implemented.

jreback commented on doc/source/whatsnew/v1.0.0.rst:
```diff
@@ -240,6 +240,7 @@ Other
 - Trying to set the ``display.precision``, ``display.max_rows`` or ``display.max_columns`` using :meth:`set_option` to anything but a ``None`` or a positive int will raise a ``ValueError`` (:issue:`23348`)
 - Using :meth:`DataFrame.replace` with overlapping keys in a nested dictionary will no longer raise, now matching the behavior of a flat dictionary (:issue:`27660`)
 - :meth:`DataFrame.to_csv` and :meth:`Series.to_csv` now support dicts as ``compression`` argument with key ``'method'`` being the compression method and others as additional compression options when the compression method is ``'zip'``. (:issue:`26023`)
+- Bug in :meth:`Series.diff` where a boolean series would incorrectly raise a TypeError (the - operator is deprecated) (:issue:`17294`)
```
use double backticks around ``TypeError``, I would remove the ``(the - operator is deprecated)``, not obvious what this is in reference to a casual reader

jreback reviewed (commented):
looks fine, small doc comment, CI is currently failing however, so wait a little for that to be updated

Unprocessable committed:
Changing bug description

Changing the description again. The numpy version and 'the - operator is deprecated' are now both removed.

jreback commented on pandas/tests/series/test_analytics.py:
```diff
@@ -237,6 +238,67 @@ def test_npdiff(self):
+        s = Series(input)
+        result = s.diff()
+        expected = Series(output)
+        assert_series_equal(result, expected)
```
is x - x.shift() also true here (as it is for obj_diff)? can you add that assert as well (do another result= and expected=), otherwise this PR lgtm.

Unprocessable replied:
No, it is not and will not with the current implementation. @TomAugspurger [specifically wanted a boolean Series](https://github.com/pandas-dev/pandas/pull/28251#discussion_r320818052) back instead of a list of 1,0,-1. As shown in the quote below.

> I'm having a hard time following, but IMO these two should have the same output
> 
> ```python
> In [3]: pd.Series([True, False, True]).diff()
> Out[3]:
> 0     NaN
> 1    True
> 2    True
> dtype: object
> 
> In [4]: pd.Series([True, False, True]).astype(object).diff()
> Out[4]:
> 0    NaN
> 1     -1
> 2      1
> dtype: object
> ```
> 
> I think both of those should be `Out[3]`, the same result as `np.diff` with the missing values handled appropriately.

jreback replied:
ok fair enough, do we have a followup issue on this? @TomAugspurger ? (which likely will block on a real BoolType EA)

jreback commented:
thanks @Unprocessable 

sorry this took a while. sometimes PRs trigger a lot of discussion about the original issue (which is what happened here).

Unprocessable commented:
No worries, it is good to see a simple question/PR causing a review of the core code. At least it means it was due for some commenting. This is also my first contribution to an open source project, so there is that. Thanks to all for keeping it updated and working!
